stream-chat-android-client/src/main/java/io/getstream/chat/android/client/call/Call.kt

```diff
@@ -1,23 +1,61 @@
 package io.getstream.chat.android.client.call
 
-import androidx.annotation.UiThread
 import androidx.annotation.WorkerThread
-import io.getstream.chat.android.client.errors.ChatError
 import io.getstream.chat.android.client.utils.Result
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.resume
 
+/**
+ * A pending operation waiting to be ran.
+ *
+ * There are several ways to run a [Call]:
+ * - [execute]: synchronous, blocking
+ * - [enqueue]: async, callback based
+ * - [await]: async, suspending call
+ *
+ * Running a [Call] more than once results in undefined behaviour.
+ */
 public interface Call<T : Any> {
-
+    /**
+     * Executes the call synchronously, in a blocking way. Only call this from a background thread.
+     */
     @WorkerThread
     public fun execute(): Result<T>
 
-    @UiThread
+    /**
+     * Executes the call asynchronously, on a background thread. Safe to call from the main
+     * thread. The [callback] will always be invoked on the main thread.
+     */
     public fun enqueue(callback: (Result<T>) -> Unit = {})
 
+    /**
+     * Cancels the execution of the call, if cancellation is supported for the operation.
+     *
+     * Note that calls can not be cancelled when running them with [execute].
+     */
     public fun cancel()
+}
+
+/**
+ * Awaits the result of [this] Call in a suspending way, asynchronously.
+ * Safe to call from any CoroutineContext.
+ *
+ * Does not throw exceptions. Any errors will be wrapped in the [Result] that's returned.
+ */
+public suspend fun <T : Any> Call<T>.await(): Result<T> = suspendCancellableCoroutine { continuation ->
+    this.enqueue { result ->
+        continuation.resume(result)
+    }
+
+    continuation.invokeOnCancellation {
+        this.cancel()
+    }
+}
+
+internal fun <T : Any, K : Any> Call<T>.map(mapper: (T) -> K): Call<K> {
+    return MapCall(this, mapper)
+}
 
-    public fun <K : Any> map(mapper: (T) -> K): Call<K>
-    public fun onError(handler: (ChatError) -> Unit): Call<T>
-    public fun onSuccess(handler: (T) -> Unit): Call<T>
-    public fun <K : Any> zipWith(call: Call<K>): Call<Pair<T, K>>
-    public fun <C : Any, B : Any> zipWith(callK: Call<C>, callP: Call<B>): Call<Triple<T, C, B>>
```

Comment:
Half of these were unused, the rest were made internal and moved into extension functions. Clients should never use these in their code, they should just run `Call` instances in one of three ways.